### PR TITLE
Fixing a pathological case for slow tokenizers

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes issue found here for slow tokenizers:

https://github.com/huggingface/tokenizers/issues/848

When using arbitrary tokens, a bug could occur where we would seed the new character, even when the lookahead attempted to skip over that part of the text. Thus we could have an extra match that didn't fit.

Other bug was during the lookahead we could undermatch (since we were iterating on new characters before checking termination).

The test added covers both cases.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -190,6 +195,10 @@ def split(self, text: str) -> List[str]:
                         # End lookahead
 
                     # Storing and resetting
+                    if offsets and start < offsets[-1]:
+                        import ipdb
+
+                        ipdb.set_trace()
```
Comment (sgugger) [inline src/transformers/tokenization_utils.py:201]: I don't think this is supposed to be here.

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

File: src/transformers/tokenization_utils.py (hunk 1 above)
Comment (Narsil) [inline src/transformers/tokenization_utils.py:201]:
```suggestion
```

Comment (Narsil) [reply src/transformers/tokenization_utils.py]: No it wasn't. :)

**Pull request merged**